scrapi/harvesters/noaa_nodc.py

```diff
@@ -0,0 +1,226 @@
+            keywords.extend(keyword_group.xpath('./gmd:keyword/gmx:Anchor/node()', namespaces=keyword_group.nsmap))
+    return keywords
+
+
+class NODCHarvester(XMLHarvester):
+    short_name = 'noaa_nodc'
+    long_name = 'National Oceanographic Data Center'
+    url = 'https://www.nodc.noaa.gov/'
+
+    DEFAULT_ENCODING = 'UTF-8'
+    record_encoding = None
+
+    canonical_base_url = 'http://data.nodc.noaa.gov/cgi-bin/iso?id={}'
+
+    namespaces = {
+        'gco': 'http://www.isotc211.org/2005/gco',
+        'gmd': 'http://www.isotc211.org/2005/gmd',
+        'gmi': 'http://www.isotc211.org/2005/gmi',
+        'gml': 'http://www.opengis.net/gml/3.2',
+        'gmx': 'http://www.isotc211.org/2005/gmx',
+        'gsr': 'http://www.isotc211.org/2005/gsr',
+        'gss': 'http://www.isotc211.org/2005/gss',
+        'gts': 'http://www.isotc211.org/2005/gts',
+        'srv': 'http://www.isotc211.org/2005/srv',
+        'xlink': 'http://www.w3.org/1999/xlink',
+        'xsi': 'http://www.w3.org/2001/XMLSchema-instance',
+        # xsi:schemaLocation="http://www.ngdc.noaa.gov/metadata/published/xsd/schema.xsd",
+    }
+
+    @property
+    def schema(self):
+        id_stanza = './gmd:identificationInfo/gmd:MD_DataIdentification/'
+        cite_stanza = id_stanza + 'gmd:citation/gmd:CI_Citation/'
+        return {
+            'title': (cite_stanza + 'gmd:title', compose(xml_text_only, single_result)),
+            'description': (id_stanza + 'gmd:abstract', compose(xml_text_only, single_result)),
+            'contributors': (cite_stanza + 'gmd:citedResponsibleParty/gmd:CI_ResponsibleParty', compose(parse_contributors, filter_to_contributors)),
+            'uris': {
+                'canonicalUri': (
+                    './gmd:fileIdentifier',
+                    compose(lambda x: str(self.canonical_base_url).format(x), xml_text_only, single_result)
+                ),
+            },
+            'publisher': (
+                cite_stanza + 'gmd:citedResponsibleParty/gmd:CI_ResponsibleParty',
+                compose(extract_organization, single_result, filter_to_publishers),
+            ),
+            'providerUpdatedDateTime': ('./gmd:dateStamp/gco:DateTime/node()', compose(date_formatter, single_result)),
+            'languages': ('./gmd:language/gmd:LanguageCode', compose(language_codes, xml_text_only_list, coerce_to_list)),
+            'subjects': (id_stanza + 'gmd:descriptiveKeywords/gmd:MD_Keywords', lambda x: filter_keywords(x)),
+        }
+
+    search_base_url = 'http://data.nodc.noaa.gov/cgi-bin/oai-pmh?verb=ListRecords'
+    oai_ns = {
+        'oai_dc': 'http://www.openarchives.org/OAI/2.0/',
+        'dc': 'http://purl.org/dc/elements/1.1/',
+    }
+
+    def query_by_date(self, start_date, end_date):
+        '''Use OAI-PMH interface to get a list of dataset ids for the given date range'''
+        search_url_end = '&metadataPrefix=oai_dc&from=' + start_date + '&until=' + end_date
```

Comment:
see previous comment on string formatting. Alternatively, look at the ```furl``` library. (this isn't that important though)